### PR TITLE
[NestedTensor] Add a NestedTensor Readme

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
This PR adds a NestedTensor Readme which explains the code structure and will hopefully serve as a reference point for new contributors, especially if they would like to implement a NestedTensor kernel implementation.

cc @cpuhrsch @jbschlosser @bhosmer @mikaylagawarecki

### CODE REVIEW

File: aten/src/ATen/native/nested/README.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+5. Construct a new NestedTensor from the output of f and the sizes of the input NestedTensor.
+
+There are also important functions that under certain conditions of regularity they can be implemented effictevly by acccesing the underlying buffer and viewing it in a special manner. For a good example of this see the implementation of `linear` in [NestedTensorTransformerFunctions.cpp](NestedTensorTransformerFunctions.cpp).
+
+The second class of functions can't be efficeintly implemented by viewing the NestedTensor as a dense tensor. An example of this is `softmax_nested`. The implementation of this function can be found in [NestedTensorMath.cpp](NestedTensorMath.cpp). This can occur when there are strict problem boundaries that occur at the tensor_component level. For example, when computing the softmax of a NestedTensor we need to compute the softmax of each tensor_component individually. This is because the softmax function is not defined on tensors of different shapes. In this case we can't view the NestedTensor as a dense tensor and apply the softmax function to the entire tensor at once. Instead we need to apply the softmax function to each tensor_component individually.
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/README.md:51]: nit: efficiently

**Review by mikaylagawarecki — APPROVED** [review]
This looks great! :) just some tiny nits, could you also link this over here in the docs, there was a  link pointing to an empty wiki previously from 1.13 that I had added https://github.com/pytorch/pytorch/blob/f571ae4fdb23ca2d22b7bba8a3b382e54fba4bb8/docs/source/nested.rst?plain=1#L25-L27

**Comment by drisspg** [comment]
> This looks great! :) just some tiny nits, could you also link this over here in the docs, there was a link pointing to an empty wiki previously from 1.13 that I had added
> 
> https://github.com/pytorch/pytorch/blob/f571ae4fdb23ca2d22b7bba8a3b382e54fba4bb8/docs/source/nested.rst?plain=1#L25-L27

I was thinking about this, do you think that we should add another level of misdirection -> point the wiki to the readme. Since I think the wiki link can/maybe should be thrown in the warn once for users that encounter problems. then if they want to contribute we can point to the readme

File: aten/src/ATen/native/nested/README.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+- nested_size_tensor_: 2d tensor of n_tensor_components x n_dims
+- nested_stride_tensor_: 2d tensor of n_tensor_components x n_dims
+- storage_offsets_: 1d tensor of offsets corresponding to the start position of each tensor component
+- storage_: The storage object that contains the flattened tensor_components (defined on c10::TensorImp)
+
+NestedTensors inherit from c10::TensorImpl whose definition can be found here: [NestedTensorImpl.h](../../NestedTensorImpl.h).
+
+When constructing a NestedTensor in C++ you will likely not be using the NestedTensorImpl constructor directly but using the `wrap_buffer` function defined in [NestedTensorUtils.h](NestedTensorUtils.h). This is a thin wrapper around the NestedTensorImpl constructor that ensures that the input tensor is contiguous. This is because when constructing a NestedTensor from a dense tensor we create a shallow copy of the input's Storage object and the if the input tensor did not satisfy `tensor.numel() == tensor.storage.numel()` this could lead to undefined behavior.
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/README.md:15]: nit: "and the if the input tensor did"

File: aten/src/ATen/native/nested/README.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+##  Code Structure
+
+The NestedTensor code is split into two parts: the C++ code and the Python code. The C++ code is located in [aten/src/ATen/native/nested](.) and the Python code is located in [torch/nested/__init__.py](/torch/nested/__init__.py). The C++ code is split into the following files:
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/README.md:19]: nit: (.)

File: aten/src/ATen/native/nested/README.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+5. Construct a new NestedTensor from the output of f and the sizes of the input NestedTensor.
+
+There are also important functions that, under certain conditions of regularity, can be implemented effectively by accessing the underlying buffer and viewing it in a special manner. For a good example of this see the implementation of `linear` in [NestedTensorTransformerFunctions.cpp](NestedTensorTransformerFunctions.cpp).
+
+The second class of functions can't be efficiently implemented by viewing the NestedTensor as a dense tensor. An example of this is `softmax_nested`. The implementation of this function can be found in [NestedTensorMath.cpp](NestedTensorMath.cpp). This can occur when there are strict problem boundaries that occur at the tensor_component level. For example, when computing the softmax of a NestedTensor we need to compute the softmax of each tensor_component individually. This is because the softmax function is not defined on tensors of different shapes. In this case we can't view the NestedTensor as a dense tensor and apply the softmax function to the entire tensor at once. Instead we need to apply the softmax function to each tensor_component individually.
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/README.md:52]: I don't think this is necessarily true if the last dimension of the NestedTensor is regular (same for any other normalization)? [softmax](https://pytorch.org/docs/master/generated/torch.nn.functional.softmax.html?highlight=softmax#torch.nn.functional.softmax) is applied at the per-vector level. The reason it's an issue for us right now, is because we apply softmax as part of SDPA's math decomposition where the matmul between query and key cause irregular last dimensions.

File: aten/src/ATen/native/nested/README.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+The problem with this though is that iterating over a potentially large number of tensor_components and launching individual cuda kernels for each one is very inefficient. Ideally we would launch one kernel that operates on all the tensor_components in parallel.
+
+If performance is not your main concern and you would like to enable coverage the function `map_nested_tensor` can be found in [NestedTensorUtils.h](NestedTensorUtils.h). This function iterates over tensor components to applying a function to each tensor_component individually. Be warned though that this is likely very inefficient, especially for cuda implementations of the function.
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/README.md:56]: nit: For CPU implementations this can actually serve as a pretty good baseline. If we had nested parallelism down we could offer a `parallel_map_nested_tensor`.

File: aten/src/ATen/native/nested/README.md (hunk 3 above)
Comment (drisspg) [reply aten/src/ATen/native/nested/README.md:19]: that just redirects you to the the current directory which is where the readme is defined

File: aten/src/ATen/native/nested/README.md (hunk 5 above)
Comment (drisspg) [reply aten/src/ATen/native/nested/README.md:56]: add a couple openmp parellel for and bobs your uncle. I remember Yifan was trying to get this to work at one point but was not succesful. I will note that though

File: aten/src/ATen/native/nested/README.md (hunk 4 above)
Comment (drisspg) [reply aten/src/ATen/native/nested/README.md:52]: ahhh you are right I think I let co-pilot get a little carried away. Do you think it would be fair to say:
```
The second class of functions can't be efficiently implemented by viewing the NestedTensor as a dense tensor. An example of this is `softmax_nested` of ragged dimensions. The implementation of this function can be found in [NestedTensorMath.cpp](NestedTensorMath.cpp). When computing the softmax over a ragged dimension of a NestedTensor the problem boundaries are not trivially separable, i.e. it is not possible to determine which elements belong to what tensor_component when folding the ragged dimension into another. Instead we apply the softmax function to each tensor_component individually, essentially applying a foreach loop over the individual tensors in the NestedTensor.
```

**Review by cpuhrsch — APPROVED** [review]
